packages/angular_devkit/core/src/jobs/api.ts

```diff
@@ -0,0 +1,521 @@
+ */
+export interface JobEventPong extends JobEventBase {
+  readonly kind: JobEventKind.Pong;
+}
+
+/**
+ * Generic event type.
+ */
+export type JobEvent<OutputT extends JsonValue> =
+  JobEventOnReady
+  | JobEventStart
+  | JobEventLog
+  | JobEventProgress
+  | JobEventOutput<OutputT>
+  | JobEventChannelCreate
+  | JobEventChannelMessage
+  | JobEventChannelError
+  | JobEventChannelComplete
+  | JobEventEnd
+  | JobEventPong;
+
+/**
+ * The context in which the job is run.
+ */
+export interface JobHandlerContext<
+  MinimumArgumentValueT extends JsonValue = JsonValue,
+  MinimumInputValueT extends JsonValue = JsonValue,
+  MinimumOutputValueT extends JsonValue = JsonValue,
+> {
+  readonly description: JobDescription;
+  readonly scheduler: Scheduler<MinimumArgumentValueT, MinimumInputValueT, MinimumOutputValueT>;
+
+  // In this context, JsonValue is comparable to `any`.
+  readonly dependencies: Job<JsonValue, JsonValue, JsonValue>[];
+
+  readonly rawInput: Observable<JobInput<MinimumInputValueT>>;
+}
+
+/**
+ * The Job Handler, which is a method that's executed for the job.
+ */
+export interface JobHandler<
+  ArgT extends JsonValue,
+  InputT extends JsonValue,
+  OutputT extends JsonValue,
+> {
+  (
+    argument: ArgT,
+    context: JobHandlerContext<ArgT, InputT, OutputT>,
+  ): Observable<JobEvent<OutputT>>;
+
+  jobDescription: Partial<JobDescription>;
+}
+
+
+/**
+ * The state of a job. These are changed as the job reports a new state through its events.
+ */
+export enum JobState {
+  /**
+   * The job was queued and is waiting to start.
+   */
+  Queued = 'queued',
+  /**
+   * The job description was found, its dependencies (see "Synchronizing and Dependencies")
+   * are done running, and the job's argument will be validated before the Job's code will
+   * be executed.
+   */
+  Ready = 'created',
+  /**
+   * The job has been started. The job implementation is expected to send this as soon as its
+   * work is starting.
+   */
+  Started = 'started',
+  /**
+   * The job has ended and is done running.
+   */
+  Ended = 'ended',
+  /**
+   * An error occured and the Job stopped because of internal state.
+   */
+  Errored = 'errored',
+}
+
+
+/**
+ * A Job instance, returned from scheduling a job. A Job Instance is _not_ serializable.
+ */
+export interface Job<
+  ArgumentT extends JsonValue = JsonValue,
+  InputT extends JsonValue = JsonValue,
+  OutputT extends JsonValue = JsonValue,
+> {
+  /**
+   * The unique ID of the job.
+   */
+  readonly id: symbol;
+
+  /**
+   * Description of the job.
+   */
+  readonly description: JobDescription;
+
+  /**
+   * Argument sent when scheduling the job. This is a copy of the argument.
+   */
+  readonly argument: ArgumentT;
+
+  /**
+   * The input to the job. This goes through the input channel as messages.
+   */
+  readonly input: Observer<JsonValue>;
+
+  /**
+   * Outputs of this Job.
+   */
+  readonly output: Observable<OutputT>;
```

Comment:
No they don't. You can have replay on them when you subscribe.

Reply:
I think that's a problem then. Under this model, non-trivial dependency graphs will cause jobs to lose messages in input/output/channels because wiring up things into the dependent job can only happen after the dependency is scheduled.

I think we need a dependency model that allows wiring up the channels in advance instead.

Reply:
Discussed outside of github, the only way to build a big dependency graph and wire things together in advance is to pause the scheduler, because adding a job as a dependency will subscribe to it, thus running the job.